libsnark/zk_proof_systems/plonk/plonk.tcc

```diff
@@ -0,0 +1,447 @@
+    std::vector<size_t> wire_permutation)
+{
+    assert(H_gen.size() > 0);
+    for (size_t i = 0; i < H_gen.size(); ++i) {
+        H_gen_permute[i] = H_gen[wire_permutation[i] - 1];
+    }
+}
+// compute the permutation polynomials S_sigma_1, S_sigma_2,
+// S_sigma_2 (see [GWC19], Sect. 8.1)
+template<typename FieldT>
+void plonk_compute_permutation_polynomials(
+    std::vector<polynomial<FieldT>> &S_polys,
+    std::vector<FieldT> H_gen_permute,
+    std::vector<polynomial<FieldT>> L_basis,
+    size_t num_gates)
+{
+    assert(S_polys.size() == NUM_HGEN);
+    assert(S_polys[0].size() == num_gates);
+    assert(H_gen_permute.size() == (NUM_HGEN * num_gates));
+    for (size_t i = 0; i < NUM_HGEN; ++i) {
+        typename std::vector<FieldT>::iterator begin =
+            H_gen_permute.begin() + (i * num_gates);
+        typename std::vector<FieldT>::iterator end =
+            H_gen_permute.begin() + (i * num_gates) + (num_gates);
+        std::vector<FieldT> S_points(begin, end);
+        plonk_interpolate_over_lagrange_basis<FieldT>(
+            S_points, S_polys[i], L_basis);
+    }
+}
+
+// A wrapper for multi_exp_method_BDLO12_signed() for multiplying a
+// single group element in G1 (a point on the curve) to a scalar
+// element in Fr
+template<typename ppT>
+libff::G1<ppT> plonk_exp_G1(
+    libff::G1<ppT> curve_point, libff::Fr<ppT> scalar_element)
+{
+    std::vector<libff::Fr<ppT>> scalar{scalar_element};
+    std::vector<libff::G1<ppT>> point{curve_point};
+    const size_t chunks = 1;
+    libff::G1<ppT> product = libff::multi_exp<
+        libff::G1<ppT>,
+        libff::Fr<ppT>,
+        libff::multi_exp_method_BDLO12_signed>(
+        point.begin(), point.end(), scalar.begin(), scalar.end(), chunks);
+    return product;
+}
+
+// A wrapper for multi_exp_method_BDLO12_signed() dot-product a
+// vector of group elements in G1 (curve points) with a vector of
+// scalar elements in Fr
+template<typename ppT>
+libff::G1<ppT> plonk_multi_exp_G1(
+    std::vector<libff::G1<ppT>> curve_points,
+    std::vector<libff::Fr<ppT>> scalar_elements)
+{
+    assert(curve_points.size() == scalar_elements.size());
+    const size_t chunks = 1;
+    libff::G1<ppT> product = libff::multi_exp<
+        libff::G1<ppT>,
+        libff::Fr<ppT>,
+        libff::multi_exp_method_BDLO12_signed>(
+        curve_points.begin(),
+        curve_points.end(),
+        scalar_elements.begin(),
+        scalar_elements.end(),
+        chunks);
+    return product;
+}
+
+//
+// Evaluate a polynomial F at the encrypted secret input
+// \secret^i*G_1 ie. compute f(\secret)*G1 = [f(\secret)]_i
+//
+// INPUT
+//
+// secret_powers_g1: \secret^i*G1: 0\le{i}<max_degree(Q[j]): 0\le{j}<n
+// f_poly: a polynomial
+//
+// OUTPUT
+//
+// [f_poly(\secret)]_1, 0\le 1<n : the "encrypted" evaluation of
+// the polynomial f_poly in the secret parameter \secret (the
+// toxic waste) multiplied by the group generator G_1 i.e. compute
+// f_poly(\secret)*G_1
+//
```

Comment:
Why start with an empty comment line and end with an empty comment as well?
FYI: If this is supposed to be a comment that needs to make it in the generated documentation, let's use `///` instead, and Doxygen will render them in the documentation.